pkg/operator/configobserver/oauth/observe_tokenconfig.go

```diff
@@ -0,0 +1,150 @@
+	if err != nil {
+		// If failed to read the config for access token max age field, continue execution with the
+		// hope to find the value for it in OAuth cluster config.
+		errs = append(errs, err)
+	}
+
+	oauthConfig, err := listers.OAuthLister().Get("cluster")
+	if err != nil {
+		// Failed to read OAuth cluster config.
+		if errors.IsNotFound(err) {
+			klog.Warning("oauth.config.openshift.io/cluster: not found")
+		}
+
+		if existingAccessTokenMaxAgeSeconds == 0 {
+			// As the value 0 here indicates that a read from existing config failed previously, use the default value.
+			return buildUnstructuredTokenConfig(defaultAccessTokenMaxAgeSeconds, tokenMaxAgePath), append(errs, err)
+		}
+		// Otherwise, return whatever is present in existing config.
+		return existingConfig, append(errs, err)
+	}
+
+	observedAccessTokenMaxAgeSeconds := float64(oauthConfig.Spec.TokenConfig.AccessTokenMaxAgeSeconds)
+	if observedAccessTokenMaxAgeSeconds == 0 {
+		// As the value 0 indicates that this field is not set in OAuth cluster config, use the default value.
+		observedAccessTokenMaxAgeSeconds = defaultAccessTokenMaxAgeSeconds
+	}
+
+	if existingAccessTokenMaxAgeSeconds != observedAccessTokenMaxAgeSeconds {
+		recorder.Eventf("ObserveTokenConfig", "%s changed from %d to %d", fieldAccessTokenMaxAgeSeconds,
+			existingAccessTokenMaxAgeSeconds,
+			observedAccessTokenMaxAgeSeconds)
+	}
+
+	return buildUnstructuredTokenConfig(observedAccessTokenMaxAgeSeconds, tokenMaxAgePath), errs
+}
+
+// ObserveAccessTokenInactivityTimeout returns an unstructured fragment of KubeAPIServerConfig that has access token inactivity timeout,
+// if there is a valid value for it in OAuth cluster config.
+func ObserveAccessTokenInactivityTimeout(genericlisters configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (ret map[string]interface{}, errs []error) {
+	errs = []error{}
+	observedConfig := map[string]interface{}{}
+	tokenConfigPath := []string{"oauthConfig", "tokenConfig"}
+	tokenInactivityTimeoutPath := append(tokenConfigPath, fieldAccessTokenInactivityTimeoutSeconds)
+	defer func() {
+		// Prune the observed config so that it only contains access token inactivity timeout field.
+		ret = configobserver.Pruned(ret, tokenInactivityTimeoutPath)
+	}()
+
+	listers, ok := genericlisters.(OAuthLister)
+	if !ok {
+		return existingConfig, append(errs, fmt.Errorf("failed to assert: given lister does not implement OAuth lister"))
+	}
+
+	existingAccessTokenInactivityTimeout, _, err := unstructured.NestedFieldCopy(existingConfig, tokenInactivityTimeoutPath...)
+	if err != nil {
+		// If failed to read existing config for access token inactivity timeout, continue execution with the
+		// hope to find the value for it in OAuth cluster config.
+		errs = append(errs, err)
+	}
+
+	var existingAccessTokenInactivityTimeoutFloat float64
+	if existingAccessTokenInactivityTimeout != nil {
+		if f, ok := existingAccessTokenInactivityTimeout.(float64); ok {
```

Comment:
this is a string, isn't it?

Reply:
It's an int32 in the Go type. Unmarshaling could give us a float64 or int64, depending on the library in use.

Reply:
if i understand correctly, `float64` is the default type for numbers when decoding json formatted data. It seems like Go standard library also follows the same. You meant to say if we use a library other than standard library, this value can be decoded to int64?

Anyways I think this issue will be resolved with inactivity timeout field in osin config is changed to time.Duration.

Reply:
> Anyways I think this issue will be resolved with inactivity timeout field in osin config is changed to time.Duration.

Any side-effects when doing that?

cc @stlaz

Reply:
> You meant to say if we use a library other than standard library, this value can be decoded to int64?

the yaml decode (sigs.k8s.io/yaml) produces int64 too I think.

Reply:
if you set it observed as int, the original Go decoded config will appear different in the final comparison https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/config_observer_controller.go#L110 because the decoder at https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/config_observer_controller.go#L79 decodes numbers as floats by default.

I tried to work around this when I was writing these time-related observers for the authn operator, but keeping these values as float ended up being the most sane option.

I don't think there should be side-effects to having this field as time.Duration (I got to this PR after the openshift/api one).

Note that `osin` and `kubecontrolplane` APIs need to match type if you want to be able to reuse this observer, and `kubecontrolplane` will require the field to be pointer due to https://github.com/openshift/origin/blob/295a35c33057749cc7bdee1f2d33ddaed5b77d60/vendor/k8s.io/kubernetes/pkg/kubeapiserver/authenticator/patch_authenticator.go#L60

Reply:
rounded to the next integer, but represented as float is fine with me. Make sure we don't loop. 

I wouldn't mind to see a klog.Warningf here if the observed value is not a float64.

Reply:
https://github.com/openshift/api/pull/674 converts the destination of this value to type `time.Duration`. With that, I am guessing we would not need to worry about losing the precision here?